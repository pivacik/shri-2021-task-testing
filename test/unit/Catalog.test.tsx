/**
 * @jest-environment jsdom
 */
import React from "react";

import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { Catalog } from "../../src/client/pages/Catalog";
import { initStore, productsLoaded } from "../../src/client/store";
import { CartApi, ExampleApi } from "../../src/client/api";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import { ProductShortInfo } from "../../src/common/types";
import { AxiosResponse } from "axios";
import { first } from "rxjs";

function mockResponse<T>(data: T): AxiosResponse<T, any> {
  return {
    status: 200,
    statusText: "OK",
    headers: {},
    config: {},
    data,
  };
}
describe("1", () => {
  it("В каталоге должны отображаться товары, список которых приходит с сервера", async () => {
    const api = new ExampleApi("");
    api.getProducts = () =>
      Promise.resolve<AxiosResponse<ProductShortInfo[], any>>(
        mockResponse([
          { id: 123, name: "chair", price: 200 },
          { id: 124, name: "pants", price: 1000 },
        ])
      );
    const cart = new CartApi();
    const store = initStore(api, cart);
    const component = (
      <MemoryRouter>
        <Provider store={store}>
          <Catalog />
        </Provider>
      </MemoryRouter>
    );
    const { getByText, getAllByTestId } = render(component);
    await waitForElementToBeRemoved(() => getByText(/loading/i));

    const firstProd = getAllByTestId(123)[0];
    const secondProd = getAllByTestId(124)[0];
    expect(firstProd).toMatchSnapshot();
    expect(secondProd).toMatchSnapshot();
  });
});
