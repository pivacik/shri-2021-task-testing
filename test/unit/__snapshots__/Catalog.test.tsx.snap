// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 В каталоге должны отображаться товары, список которых приходит с сервера 1`] = `
<div
  class="col-12 col-sm-6 col-md-4 col-lg-3"
  data-testid="123"
>
  <div
    class="ProductItem card w-100 mb-4"
    data-testid="123"
  >
    <img
      class="Image card-img-top"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
    />
    <div
      class="card-body"
    >
      <h5
        class="ProductItem-Name card-title"
      >
        chair
      </h5>
      <p
        class="ProductItem-Price card-text"
      >
        $
        200
      </p>
      <a
        class="ProductItem-DetailsLink card-link"
        href="/catalog/123"
      >
        Details
      </a>
    </div>
  </div>
</div>
`;

exports[`1 В каталоге должны отображаться товары, список которых приходит с сервера 2`] = `
<div
  class="col-12 col-sm-6 col-md-4 col-lg-3"
  data-testid="124"
>
  <div
    class="ProductItem card w-100 mb-4"
    data-testid="124"
  >
    <img
      class="Image card-img-top"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
    />
    <div
      class="card-body"
    >
      <h5
        class="ProductItem-Name card-title"
      >
        pants
      </h5>
      <p
        class="ProductItem-Price card-text"
      >
        $
        1000
      </p>
      <a
        class="ProductItem-DetailsLink card-link"
        href="/catalog/124"
      >
        Details
      </a>
    </div>
  </div>
</div>
`;
